Update Internals::touchEventRectsForEvent to use new touch event region codepath

nmoucht opened the pull request:
#### 1ed025da72085637dbf40591545f4d29301c3234
<pre>
Update Internals::touchEventRectsForEvent to use new touch event region codepath
<a href="https://bugs.webkit.org/show_bug.cgi?id=291986">https://bugs.webkit.org/show_bug.cgi?id=291986</a>
<a href="https://rdar.apple.com/149900530">rdar://149900530</a>

Reviewed by NOBODY (OOPS!).

Update Internals::touchEventRectsForEvent to get the touch event regions from the relevant layers
rather than through the scrolling coordinator.

* LayoutTests/fast/events/touch/ios/touch-event-regions/document.html:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getEventRegions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::touchEventRectsForEventForTesting):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setEventRegion):
* Source/WebCore/rendering/EventRegion.h:
(WebCore::EventRegion::touchEventListenerRegion const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paintObject):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::getEventRegion):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::getEventRegions):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::touchEventRectsForEvent):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ed025da72085637dbf40591545f4d29301c3234

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106344 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51823 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77079 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34112 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16323 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91414 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16140 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9425 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51171 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86028 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9498 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108700 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20850 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86048 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28686 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87617 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85594 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30325 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8043 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22423 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28254 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33525 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28066 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31386 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29624 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6043,6 +6043,16 @@ void Document::updateEventRegions()
     }
 }
 
+EventTrackingRegions Document::getTouchEventRegions()
```
`Document::getTouchEventRegions() const`.

We generally don't say "get". Just `touchEventRegions()` should be fine.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/EventRegion.h:
```diff
@@ -168,6 +168,10 @@ class EventRegion {
 #endif
 
+#if ENABLE(TOUCH_EVENT_REGIONS)
+    EventTrackingRegions touchEventListenerRegion() const { return m_touchEventListenerRegion; }
```
The mismatch between "region" and regions" hurts my brain. Choose one!

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4397,6 +4397,16 @@ void RenderLayer::collectEventRegionForFragments(const LayerFragments& layerFrag
     }
 }
 
+EventTrackingRegions RenderLayer::getTouchEventRegion()
```
No "get". Can be const. Should be plural.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1003,6 +1003,21 @@ void RenderLayerCompositor::updateEventRegions()
+    Vector<RenderLayer*> layers = { m_renderView.layer() };
+    while (!layers.isEmpty()) {
+        if (auto* layer = layers.takeLast()) {
+            eventRegion.unite(layer->getTouchEventRegion());
```
I'm not sure this is right; isn't this uniting regions in different coordinate systems?

nmoucht replied:
Right yeah we should collect them separately.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.h:
```diff
@@ -209,6 +209,7 @@ class RenderLayerCompositor final : public GraphicsLayerClient, public CanMakeCh
     // Update event regions, which only needs to happen once per rendering update.
     void updateEventRegions();
     void updateEventRegionsRecursive(RenderLayer&);
+    EventTrackingRegions getTouchEventRegions();
```
If this is just for testing, name it thusly.

nmoucht replied:
Fixed.

nmoucht committed:
Update Internals::touchEventRectsForEvent to use new touch event region codepath
https://bugs.webkit.org/show_bug.cgi?id=291986
rdar://149900530

Reviewed by NOBODY (OOPS!).

Update Internals::touchEventRectsForEvent to get the touch event regions from the relevant layers
rather than through the scrolling coordinator.

* LayoutTests/fast/events/touch/ios/touch-event-regions/document.html:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getEventRegions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::touchEventRectsForEventForTesting):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setEventRegion):
* Source/WebCore/rendering/EventRegion.h:
(WebCore::EventRegion::touchEventListenerRegion const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paintObject):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::getEventRegion):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::getEventRegions):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::touchEventRectsForEvent):

smfr commented on LayoutTests/fast/events/touch/ios/touch-event-regions/active-passive-nesting.html:
```diff
@@ -28,9 +28,23 @@
+            testRunner.waitUntilDone();
+        }
+
+
```
Odd amount of whitespace in this test.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1003,6 +1003,21 @@ void RenderLayerCompositor::updateEventRegions()
+    Vector<CheckedPtr<RenderLayer>> layers = { m_renderView.layer() };
+    while (!layers.isEmpty()) {
+        if (CheckedPtr layer = layers.takeLast()) {
+            eventRegions.append(layer->touchEventRegionsForTesting());
```
This is still collecting a vector of regions in different coordinate systems.

I think the better approach here is to abandon `touchEventRegionsForTesting`, and instead have tests dump layer trees with the `LAYER_TREE_INCLUDES_EVENT_REGION` flag, and make sure your touch event regions show there.
